cmd/earth/main.go

```diff
@@ -1460,17 +1462,78 @@ func (app *earthApp) newBuildkitdClient(ctx context.Context, opts ...client.Clie
 	return bkClient, nil
 }
 
+func detectCI() string {
+	for k, v := range map[string]string{
+		"GITHUB_WORKFLOW": "github-actions",
+		"CIRCLECI":        "circle-ci",
+		"JENKINS_HOME":    "jenkins",
+		"BUILDKITE":       "buildkite",
+		"DRONE_BRANCH":    "drone",
+		"TRAVIS":          "travis",
+	} {
+		if _, ok := os.LookupEnv(k); ok {
+			return v
+		}
+	}
+
+	// default catch-all
+	if v, ok := os.LookupEnv("CI"); ok {
+		if strings.ToLower(v) == "true" {
+			return "unknown"
+		}
+		return v
+	}
+
+	return "false"
+}
+
+func getRepo() string {
+	cmd := exec.Command("git", "config", "--get", "remote.origin.url")
+	out, err := cmd.Output()
+	if err == nil {
+		return string(out)
+	}
+
+	for _, k := range []string{
+		"GITHUB_REPOSITORY",
+		"CIRCLE_PROJECT_REPONAME",
+		"GIT_URL",
+		"BUILDKITE_REPO",
+		"DRONE_REPO",
+		"TRAVIS_REPO_SLUG",
+	} {
+		if v, ok := os.LookupEnv(k); ok {
+			return v
+		}
+	}
+
+	for _, e := range os.Environ() {
+		pair := strings.SplitN(e, "=", 2)
+		if len(pair) == 2 {
+			if strings.Contains(pair[1], "git") {
+				return pair[1]
+			}
+		}
+	}
+
+	return "unknown"
+}
+
+func getRepoHash() string {
+	repo := getRepo()
+	if repo == "unknown" || repo == "" {
+		return repo
+	}
+	return fmt.Sprintf("%x", sha256.Sum256([]byte(repo)))
+}
```

Comment:
It's probably worth moving all of these into a new `analytics.go` file. `main.go` is getting pretty big.

Reply:
i was on the fence about that, but yeah that's a good idea.